File: server/src/models/AuditLog.js
const mongoose = require('mongoose');

const auditLogSchema = new mongoose.Schema(
  {
    action: {
      type: String,
      required: true,
      enum: [
        'USER_CREATED',
        'USER_UPDATED',
        'USER_DELETED',
        'ROLE_CHANGED',
        'PASSWORD_RESET',
        'REQUEST_CREATED',
        'REQUEST_APPROVED',
        'REQUEST_REJECTED',
        'REQUEST_EDITED',
        'LOGIN',
      ],
    },
    performedBy: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      required: true,
    },
    targetType: {
      type: String,
      enum: ['User', 'IdentityRequest'],
      required: true,
    },
    targetId: {
      type: mongoose.Schema.Types.ObjectId,
      required: true,
    },
    details: {
      type: mongoose.Schema.Types.Mixed,
      default: {},
    },
    ipAddress: {
      type: String,
      default: '',
    },
  },
  {
    timestamps: true,
  }
);

// Prevent modifications - audit logs are immutable
auditLogSchema.pre('findOneAndUpdate', function () {
  throw new Error('Audit logs cannot be modified');
});

auditLogSchema.pre('updateOne', function () {
  throw new Error('Audit logs cannot be modified');
});

auditLogSchema.index({ createdAt: -1 });
auditLogSchema.index({ performedBy: 1 });
auditLogSchema.index({ action: 1 });

module.exports = mongoose.model('AuditLog', auditLogSchema);

File: mobile/lib/models/audit_log.dart
class AuditLog {
  final String id;
  final String action;
  final String? performedByName;
  final String? performedByEmail;
  final String? performedByRole;
  final String targetType;
  final String targetId;
  final Map<String, dynamic> details;
  final DateTime createdAt;

  AuditLog({
    required this.id,
    required this.action,
    this.performedByName,
    this.performedByEmail,
    this.performedByRole,
    required this.targetType,
    required this.targetId,
    this.details = const {},
    required this.createdAt,
  });

  factory AuditLog.fromJson(Map<String, dynamic> json) {
    String? performedByName;
    String? performedByEmail;
    String? performedByRole;
    if (json['performedBy'] is Map) {
      performedByName = json['performedBy']['name'];
      performedByEmail = json['performedBy']['email'];
      performedByRole = json['performedBy']['role'];
    }

    return AuditLog(
      id: json['_id'] ?? '',
      action: json['action'] ?? '',
      performedByName: performedByName,
      performedByEmail: performedByEmail,
      performedByRole: performedByRole,
      targetType: json['targetType'] ?? '',
      targetId: json['targetId'] ?? '',
      details: json['details'] is Map<String, dynamic>
          ? json['details']
          : {},
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'])
          : DateTime.now(),
    );
  }

  String get actionDisplay {
    const labels = {
      'USER_CREATED': 'User Created',
      'USER_UPDATED': 'User Updated',
      'USER_DELETED': 'User Deleted',
      'ROLE_CHANGED': 'Role Changed',
      'PASSWORD_RESET': 'Password Reset',
      'REQUEST_CREATED': 'Request Created',
      'REQUEST_APPROVED': 'Request Approved',
      'REQUEST_REJECTED': 'Request Rejected',
      'LOGIN': 'Login',
    };
    return labels[action] ?? action;
  }
}
